src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs

```diff
@@ -0,0 +1,117 @@
+            Debug.Assert(length != 0 && length != this.Length, "Caller should handle Length boundary conditions.");
+
+            Utf8String newString = FastAllocate(length);
+            Buffer.Memmove(ref newString.DangerousGetMutableReference(), ref this.DangerousGetMutableReference(startIndex), (uint)length);
+            return newString;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Utf8String Substring(Index startIndex)
+        {
+            int actualIndex = startIndex.GetOffset(Length);
+            return Substring(actualIndex);
+        }
+
+        public Utf8String Substring(int startIndex)
+        {
+            if ((uint)startIndex > (uint)this.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startIndex);
+            }
+
+            // Optimizations: since instances are immutable, we can return 'this' or the known
+            // Empty instance if the caller passed us a startIndex at the string boundary.
+
+            if (startIndex == 0)
+            {
+                return this;
+            }
+
+            if (startIndex == Length)
+            {
+                return Empty;
+            }
+
+            return InternalSubstring(startIndex, Length - startIndex);
+        }
+
+        public Utf8String Substring(int startIndex, int length)
+        {
+            ValidateStartIndexAndLength(startIndex, length);
+
+            // Optimizations: since instances are immutable, we can return 'this' or the known
+            // Empty instance if the caller passed us a startIndex at the string boundary.
+
+            if (length == 0)
+            {
+                return Empty;
+            }
+
+            if (length == this.Length)
+            {
+                return this;
+            }
+
+            return InternalSubstring(startIndex, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Utf8String Substring(Range range)
+        {
+            (int start, int length) = range.GetOffsetAndLength(Length);
+            return Substring(start, length);
+        }
+
+        [StackTraceHidden]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ValidateStartIndexAndLength(int startIndex, int length)
+        {
+#if BIT64
+            // See comment in Span<T>.Slice for how this works.
+            if ((ulong)(uint)startIndex + (ulong)(uint)length > (ulong)(uint)this.Length)
+                ValidateStartIndexAndLength_Throw(startIndex, length);
+#else
+            if ((uint)startIndex > (uint)this.Length || (uint)length > (uint)(this.Length - startIndex))
+                ValidateStartIndexAndLength_Throw(startIndex, length);
+#endif
+        }
+
+        [StackTraceHidden]
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
No-inline attribute's side-effects is that the method is no longer recognized by JIT as never returning and thus its callsites are not marked as cold code and are also emitted in a way that allows returning (registers preserved, reversed branch condition, etc).
Could be solved by not using `ThrowHelper` or by adding a `throw null;´ to the end, but no-inline should be removed either way.
The "proper/final" solution could be the never/bottom type: https://github.com/dotnet/csharplang/issues/538

Reply:
Should still add a `throw null` at the end of the method so that is properly detected as never returning (JIT detects it by observing no `ret` instructions in the method body).
Compare [Substring w/o and w/ throw null](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKB3wIDoAlAVwDs4BLVAU2oGEB7VABxYAbdgCcAymIBuLAMbtkpMuUwBmfACZcAQVwBvcriO4WzXABl2jAOZwAFkuO5Dxtbh7AAVu1lxc4+mBCbAAKUz9kOERROABJRgATdjAYEzMRG3sAShcjAzInJwA1RCEWBMQ4dnEomPiksG1Ey0y7EMjouMTk1IzbOyylXMKAehHcAHl+VlQWAC9Klh5GZBBcZFN5NI7GeWRcaPYTVFR6KOARVIB3I9lERnwAdlwAcnsWZBf3UVx7I4BrRg8K6MYZOMa4ACiAjgAE9tlFdkcWAAzX52W6lEQ/fgoZDsBK4ej7RDrWpdBoHPx/Kh4YA8JgVUSw6jkMHGVG4EJ9ey4AC8fNw2ByBUKeXZYswzwARNLHGLcABfNmisWc7lWfr8wXvZDUFr9EUK/QSwpS9EfeVi5XKVVm57xKqiRilAJBAihDp1bopXA8gZWpUqpwAbQAsux7DwErEBEIQhGozG41NWMs9QA5HjxMqMUzWLIAXWG/FELCklSOukdYhdQjdwTCZi9FJ6aT8/qy/IAfLhGPQhEIhnbw5G7NHY/x44nx8mp6mlitqNprNZRAoNlJ2DnTPmiyWyxWqvgoLgSmUKlUap16skmgkDfYmxFybffeE/ZrssN8saIZJbj4DgzFMfxcUYAAeAAVbtqHEMothRHgfnHK4LX2K5kP+PVTXVEIB2WAt8PCLIWzfXAAGouQImwsmI5gsn9XBe3woRCLo+gSN1fUvwGU1ilKcpK2vb0GnvR87AAfSguxRGBdpXx9XpeMGYYbQPctKxPM9BMvapFLE5peOk2T5I/MilPbT9WiNYxfwVPDOIYizKRYpy4CybiJNssV7ONIwZLkq5tFEax6GAuAJjOCYUVoe5rHYSEwHkaZFxCR5VLtJwbWNdghHxU0/P8wLgRCsKIqiyLYvixLkvYVLlhCAAOTLjRy4x1LIABIUtNOPQgIG0krgtC8KrEi6LqpsWqUrTRhn1wKQuz5Xt7CCvsByHchOrcTAtAAIRNO0PwkwNhjcDxvF8fxAkbcyDLbD9Ox/U1zyEq8HsaIzWgUm9LM7Yc/3GBdZgWObVnWTZkRWRE9gOddjlOc5LlwG5cDuB5zTeOwPi+ZD0QBIEQVNCFoWmeFTB2LZORpO5BzEXBcWQfFCWJA4yT+ylKgJ2lcHpRlohZYMHLRDVWm1IUfMKIrjXNWVA2y4W1VFpiBXQnibMK/jXGeXUFY6pX7VwGtnVdW6PV+0S2wBtTDdHJNJ2nMcJxTBqlyzHc81o4s7V6o8jkOk26wbC37s5x70hUnsNsHQHjHt2dHYTZ25yEBd02XVd12Z8tt0YXM9x9pw/a0zBTzevSRNbL6H14haXIjjsVJerLjH/dhAJOca0jA+5oNg+C5COJCUOBdDUawnDW6MRy2No+iPIbsBKOoueiPcxjeOY1f2IXzycb1bztaMCvhM+8TjOGy3q+UmzA064vD1L8vdLP8Oa4kkygvrz7Uie5u7Qyw5KLPeS9t57y8gA40QCxTDTKmNZglUYpxWmklWaaUMr6yMO1MUeUCrTyOv5YwcDRoVUmighKaD6pzWaq1BUODChrTHv2WOttbSPz6kcAaQ1TIjXKuNJBU1KF1TdgtJa0cmFoRYVtMgiogA).

Reply:
Or just use regular `throw` without the throw helpers goop. It will solve all the problems discussed here.